resolwe/flow/managers/consumer.py

```diff
@@ -1,24 +1,96 @@
 """Manager Channels consumer."""
-from channels.generic import BaseConsumer
+import asyncio
 
-from . import manager, state
+import async_timeout
+from asgiref.sync import async_to_sync
+from channels.consumer import AsyncConsumer
+from channels.layers import get_channel_layer
+from channels.testing import ApplicationCommunicator
 
+from . import state
 
-class ManagerConsumer(BaseConsumer):
+
+class ManagerConsumer(AsyncConsumer):
     """Channels consumer for handling manager events."""
 
-    @classmethod
-    def channel_names(cls):
-        """Names of channels to listen for messages on."""
-        return {state.MANAGER_CONTROL_CHANNEL}
+    def __init__(self, *args, **kwargs):
+        """Initialize a consumer instance with the given manager."""
+        # This import is local in order to avoid startup import loops.
+        from . import manager
+        self.manager = manager
+        super().__init__(*args, **kwargs)
+
+    async def control_event(self, message):
+        """Forward control events to the manager dispatcher."""
+        try:
+            await self.manager.handle_control_event(message['content'])
+        finally:
+            await self.manager.sync_counter.dec()
+
+    @staticmethod
+    async def send_event(message):
+        """Construct a Channels event packet with the given message.
+
+        :param message: The message to send to the manager workers.
+        """
+        packet = {
+            'type': 'control_event',  # This is used as the method name in the consumer.
+            'content': message,
+        }
+        await get_channel_layer().send(state.MANAGER_CONTROL_CHANNEL, packet)
+
+    @staticmethod
+    async def flush_queue():
```

Comment:
These static methods are in the consumer but they are called from the manager? And the consumer calls the manager, getting us a full circle. Why aren't these methods in the manager if they are only used there?

Reply:
Also in response to the previous comment.

I think it makes sense to segregate everything to do with Channels into the consumer. This way, the main manager code can be mostly about what the manager actually is, along with some additional utility functions (because we don't want the world to know about the consumer at all). The consumer is a specific implementation for outside infrastructure (Channels in this case) on which the manager sits, so it's sensible to make it handle everything to do with that infrastructure.

The manager itself used to have lots of test-specific functionality and that code wasn't in the tests either. Now some of it moved to the consumer, because it makes more sense there (and also the dispatcher is getting rather large, that's why I tried keeping as much Channels code out of it as possible).

As for the circle: it could be broken by moving all the low level Channels management code into the manager, but I think that would make everything much more ugly. The circle is controlled in that the manager only calls consumer methods which are call graph leaves. I could make those methods top-level instead of static.